### PR TITLE
Live TV improvements

### PR DESCRIPTION
**Pull request by MrChip53**
### Changes
- Added channel name to info bar for programs with no data.
- Ignore programs that fill time slots that have already been filled by a previous program.
- Change live tv guide from 6 hours to 9 hours of programming to be consistent with non live guide.
- Channel surf using channel headers in live guide
- Long/short press in non-live guide and live guide
- Added full back button handling properly
- Split no program data into 30 minute slots for easier side scrolling
- Fixed favorite programs not sorting to the top when the preference is checked

### Known Issues
- ~~When watching live TV the live guide is set to force refresh every 30 minutes, it appears when it is supposed to refresh the loading spinner will not disappear.~~

### Possible Issues(I can't reproduce)
- None

### To do (hopefully):
- [x] Address issue #284 
~~1. Look into channel sorting~~ **My guide data is so bad I can't get good results no matter what here**
~~2. Lazy load channels~~ **Will work on in different commit**
~~3. If program is in a time slot that conflicted with another program then highlight the background to signify we don't know if this is what is actually playing~~ **Scrapped for now**

### Finished:
- [x] Replace \<No Prorgam Data\> with resource string
- [x] Skip program data if another program has filled the time slot
- [x] Channel headers can be used to surf in guide
- [x] Short press to play channel, long press for detail screen
- [x] Split no program data in 30 minute slots

**Maybe implement in the far off future**
- Do not refresh entire player when switching channels by channel header, this appears to only happen if channel doesn't work with the player(libVlc or exoplayer) currently being used
- Split \<No Program Data\> shows into 30 minute segments for the filler slot

### CODE REVIEW

File: app/src/main/java/org/jellyfin/androidtv/ui/GuideChannelHeader.java [hunk 1]
```diff
@@ -45,14 +48,39 @@ private void initComponent(Context context, ChannelInfoDto channel) {
...
+        super.onFocusChanged(gainFocus, direction, previouslyFocusedRect);
+
+        if (gainFocus) {
+            setBackgroundColor(Utils.getThemeColor(getContext(), android.R.attr.colorAccent));
+
+            mActivity.setSelectedProgram(this);
+        } else {
+            setBackground(getResources().getDrawable(R.drawable.light_border));
+        }
+//        TvApp.getApplication().getLogger().Debug("Focus on " + mProgram.getName() + " was " + (gainFocus ? "gained" : "lost"));
```
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/ui/GuideChannelHeader.java:83]: Either remove, uncomment or add a comment explaining why this line is commented to avoid confusion for other devs

Comment (MrChip53) [reply app/src/main/java/org/jellyfin/androidtv/ui/GuideChannelHeader.java:83]: I'll pull it. That line was already there. I didn't know why myself.

File: app/src/main/java/org/jellyfin/androidtv/ui/livetv/LiveTvGuideActivity.java [hunk 2]
```diff
@@ -767,26 +797,36 @@ protected void onPostExecute(Void aVoid) {
...
-            cell.setLast();
-            cell.setFirst();
+            int minutes = ((Long)((mCurrentLocalGuideEnd - mCurrentLocalGuideStart) / 60000)).intValue();
+            int slot = 0;
+            do {
+                BaseItemDto empty = new BaseItemDto();
+                empty.setName("  " + getString(R.string.no_program_data));
+                empty.setChannelId(channelId);
+                empty.setStartDate(TimeUtils.convertToUtcDate(new Date(mCurrentLocalGuideStart + ((30*slot) * 60000))));
+                empty.setEndDate(TimeUtils.convertToUtcDate(new Date(mCurrentLocalGuideStart + ((30+(slot+1)) * 60000))));
```
Comment (thornbill) [inline app/src/main/java/org/jellyfin/androidtv/ui/livetv/LiveTvGuideActivity.java:818]: Shouldn’t this be `30*(slot+1)`?

File: app/src/main/java/org/jellyfin/androidtv/ui/livetv/LiveTvGuideActivity.java [hunk 3]
```diff
@@ -767,26 +797,36 @@ protected void onPostExecute(Void aVoid) {
...
+                cell.setId(currentCellId++);
+                cell.setLayoutParams(new ViewGroup.LayoutParams(30 * PIXELS_PER_MINUTE, ROW_HEIGHT));
+                cell.setFocusable(true);
+                programRow.addView(cell);
+                if (slot == 0)
+                    cell.setFirst();
+                if (slot == (minutes / 30) - 1)
+                    cell.setLast();
+                slot++;
+            } while((30*slot)*60000 < minutes*60000);
```
Comment (thornbill) [inline app/src/main/java/org/jellyfin/androidtv/ui/livetv/LiveTvGuideActivity.java:829]: It shouldn’t be necessary to multiply both sides by 60000 here.

File: app/src/main/java/org/jellyfin/androidtv/ui/playback/CustomPlaybackOverlayFragment.java [hunk 4]
```diff
@@ -842,19 +910,35 @@ protected void onPostExecute(Void aVoid) {
...
+
+            int minutes = ((Long)((mCurrentLocalGuideEnd - mCurrentLocalGuideStart) / 60000)).intValue();
+            int slot = 0;
+
+            do {
+                BaseItemDto empty = new BaseItemDto();
+                empty.setName("  " + getString(R.string.no_program_data));
+                empty.setChannelId(channelId);
+                empty.setStartDate(TimeUtils.convertToUtcDate(new Date(mCurrentLocalGuideStart + ((30*slot) * 60000))));
+                empty.setEndDate(TimeUtils.convertToUtcDate(new Date(mCurrentLocalGuideStart + ((30+(slot+1)) * 60000))));
```
Comment (thornbill) [inline app/src/main/java/org/jellyfin/androidtv/ui/playback/CustomPlaybackOverlayFragment.java:929]: Same here

File: app/src/main/java/org/jellyfin/androidtv/ui/playback/CustomPlaybackOverlayFragment.java [hunk 5]
```diff
@@ -842,19 +910,35 @@ protected void onPostExecute(Void aVoid) {
...
+                cell.setId(currentCellId++);
+                cell.setLayoutParams(new ViewGroup.LayoutParams(30 * PIXELS_PER_MINUTE, LiveTvGuideActivity.ROW_HEIGHT));
+                cell.setFocusable(true);
+                programRow.addView(cell);
+                if (slot == 0)
+                    cell.setFirst();
+                if (slot == (minutes / 30) - 1)
+                    cell.setLast();
+                slot++;
+            } while((30*slot)*60000 < minutes*60000);
```
Comment (thornbill) [inline app/src/main/java/org/jellyfin/androidtv/ui/playback/CustomPlaybackOverlayFragment.java:940]: Same here

**Review by thornbill — CHANGES_REQUESTED** [review]
Looks good overall! I noticed a couple small things, but still want to test it out.

File: app/src/main/java/org/jellyfin/androidtv/ui/livetv/LiveTvGuideActivity.java (hunk 2 above)
Comment (MrChip53) [reply app/src/main/java/org/jellyfin/androidtv/ui/livetv/LiveTvGuideActivity.java]: Yes. Fixed.

File: app/src/main/java/org/jellyfin/androidtv/ui/livetv/LiveTvGuideActivity.java (hunk 3 above)
Comment (MrChip53) [reply app/src/main/java/org/jellyfin/androidtv/ui/livetv/LiveTvGuideActivity.java]: Agree. Didn't realize the redundancy. Fixed.

File: app/src/main/java/org/jellyfin/androidtv/ui/livetv/LiveTvGuideActivity.java [hunk 6]
```diff
@@ -260,7 +260,10 @@ protected void onResume() {
     }
 
     protected void doLoad() {
-        if (TvManager.shouldForceReload() || System.currentTimeMillis() > mLastLoad + 3600000) {
+        Calendar curTime = Calendar.getInstance();
+        curTime.setTime(new Date());
```
Comment (thornbill) [inline app/src/main/java/org/jellyfin/androidtv/ui/livetv/LiveTvGuideActivity.java:264]: Why did you change this from using `System.currentTimeMillis()`?

**Review by thornbill — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/jellyfin/androidtv/ui/livetv/LiveTvGuideActivity.java (hunk 6 above)
Comment (MrChip53) [reply app/src/main/java/org/jellyfin/androidtv/ui/livetv/LiveTvGuideActivity.java:264]: I changed a lot of these to use Calendar. For some reason, it appears that the calendar.getTimeMs fucntion adds some sort of zero padding and so Calendars current time would not ever equal Systems current time for me. I can look into this further and see if it was just my bad code or a real issue. The mCurrentGuide longs are set based off of Calendar variables.

**Review by thornbill — APPROVED** [review]
